quil-plotting/quil_plotting/dataframe.py

```diff
@@ -0,0 +1,346 @@
+"""dataframe module.
+-----------------
+
+This module produces a plottable dataframe with IQ values and metadata.
+
+Todo:
+- Handle runner ordering
+- Use quil waveform templates
+"""
+
+import json
+from enum import Enum
+from itertools import cycle
+from typing import Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from pyquil.quiltwaveforms import DragGaussianWaveform, ErfSquareWaveform, FlatWaveform, GaussianWaveform
+from quil.expression import Expression, PrefixExpression, PrefixOperator
+from quil.instructions import Gate, Pragma, Pulse, Qubit, ShiftPhase
+from quil.program import Program
+
+
+# color
+class Color(str, Enum):
+    GREEN = "#00b5ad"  # Teal
+    YELLOW = "#ffc504"  # Yellow
+    RED = "#ef476f"  # Magneta
+    BLUE = "#3d47d9"  # Palatinate Blue
+    GRAY = "#8a8b92"  # Gray
+    NAVY = "#0d0d36"  # Cetacean Blue
+
+
+def hex_to_rgba(hex_color: str, opacity: float = 1.0) -> str:
+    """Convert a hex color to an rgba string.
+
+    >>> hex_to_rgba("#00B5AD")
+    'rgba(0,181,173,1.0)'
+
+    :param hex_color: The hex color code.
+    :param opacity: The opacity for the rgba color.
+    """
+    hex_color = hex_color.strip("#")
+    r, g, b = tuple(int(hex_color[i : i + 2], 16) for i in (0, 2, 4))
+    return f"rgba({r},{g},{b},{opacity})"
+
+
+def rgba_to_hex(rgba: Union[np.ndarray, List[int]]) -> str:
+    """Convert an RGBA array to a hex color string.
+
+    :param rgba: Array or list containing RGB components on [0-255]. The alpha component is optional.
+    :return: Hex color string.
+    """
+    if len(rgba) == 3:
+        return f"#{rgba[0]:02x}{rgba[1]:02x}{rgba[2]:02x}"
+    elif len(rgba) == 4:
+        return f"#{rgba[0]:02x}{rgba[1]:02x}{rgba[2]:02x}{rgba[3]:02x}"
+    else:
+        raise ValueError("RGBA array must have 3 or 4 elements")
+
+
+def expand_with_source_mapping(program: Program, include_pragmas: bool = False) -> Tuple[Program, List[int]]:
```

Comment:
We should probably use the recently added [expand_calibrations_with_source_map](https://github.com/rigetti/quil-rs/blob/main/quil-py/quil/program/__init__.pyi#L140) instead of a manual implementation here.